core/src/main/java/io/grpc/internal/ManagedChannelImpl.java

```diff
@@ -1841,8 +1841,9 @@ private void handleErrorInSyncContext(Status error) {
         return;
       }
       // Apply Default Service Config if initial name resolution fails.
-      if (defaultServiceConfig != null) {
+      if (lastServiceConfig != EMPTY_SERVICE_CONFIG && defaultServiceConfig != null) {
```

Comment:
Please note now with the few additional changes suggested by Eric, we are now having to also make the unit test cover these changes viz., 
1. For testing post onError behavior: A test calling that calls onError twice. After first onError 2 trace events (failed to resolve name, applied default service config) will be added but after the 2nd onError since the default service config is not applied again, there should only be 1 additional trace event (failed to resolve name).
2. For testing onResult following by onError behavior: After an onResult with a service config that gets applied and stored in lastServiceConfig, a subsequent call to onError should result only in only trace event, for failing to resolve name.

Reply:
@kannanjgithub Fixed the UT for Point 1 in above comment

Reply:
@kannanjgithub Added the new UT for Point 2.

Reply:
I realized I made a typo in my suggesting if (lastServiceConfig != EMPTY_SERVICE_CONFIG && defaultServiceConfig != null)
It should actually be
if (lastServiceConfig == EMPTY_SERVICE_CONFIG && defaultServiceConfig != null)

Only this will implement Eric's suggestion of "If we've already updated the service config, we want to continue using it". i.e. If we updated the service config earlier from a successful onResult2 call, then lastServiceConfig would not be EMPTY_SERVICE_CONFIG. 

In your test testApplyDefaultServiceConfigIfInitialNameResolutionFails when you call resolver.listener.onError(resolutionError); with your current change
  if (lastServiceConfig != EMPTY_SERVICE_CONFIG && defaultServiceConfig != null) {
it should have normally failed this condition, and helped us realize the typo in my original suggestion.

But it did not.
This is because of something else that is going on, that is setting the lastServiceConfig to a non empty value even before your first onError call:
    In your test when you [construct](https://github.com/vinodhabib/grpc-java/blob/bc1280dd1227ffdf2ded5a8073dfe4c0fce9d411/core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java#L4483) the fake nameresolver factory, the fake name resolver immediately fakes a successful resolution call [here](https://github.com/vinodhabib/grpc-java/blob/bc1280dd1227ffdf2ded5a8073dfe4c0fce9d411/core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java#L4597) that already sets the lastServiceConfig to a non empty value.
To avoid this from happening, you can remove your first onError call and construct the FakeNameResolverFactory passing the error to its [constructor](https://github.com/vinodhabib/grpc-java/blob/bc1280dd1227ffdf2ded5a8073dfe4c0fce9d411/core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java#L4539), and it will do the onError call automatically at start, so you don't have to do the first onError explicitly in your test. 

----------------------------------------------------
Similarly, in the testTraceEventsOnResultsCallFollowedByOnErrorCallWithAppliedServiceConfig you can remove [these lines](https://github.com/vinodhabib/grpc-java/blob/defect-11040-DefaultServiceConfigFix/core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java#L4538C1-L4544C78) because the first call is already happening when the FakeNameResolver starts.
 
In addition, once you change to lastServiceConfig == EMPTY_SERVICE_CONFIG in ManagedChannelImpl, this test will also need the default service config to be set in the channel config similar to how you are already doing it in the [first test](https://github.com/vinodhabib/grpc-java/blob/defect-11040-DefaultServiceConfigFix/core/src/test/java/io/grpc/internal/ManagedChannelImplTest.java#L4489-L4493).

Reply:
In testApplyDefaultServiceConfigIfInitialNameResolutionFails I don't see the FakeNameResolverFactory constructed with the error when constructing it like I suggested.

Reply:
Fixed